src/libsync/creds/oauth.cpp

```diff
@@ -636,4 +571,89 @@ void AccountBasedOAuth::fetchWellKnown()
+                    Q_EMIT refreshError(reply->error(), reply->errorString());
+                    return;
+                } else {
+                    if (jsonParseError.error != QJsonParseError::NoError || data.isEmpty()) {
+                        // Invalid or empty JSON: Network error maybe?
+                        qCWarning(lcOauth) << "Error while refreshing the token:" << jsonParseError.errorString();
+                    } else {
+                        QString error;
+                        accessToken = getRequiredField(data, QStringLiteral("access_token"), &error).toString();
+                        if (!error.isEmpty()) {
+                            qCWarning(lcOauth) << "The reply from the server did not contain all expected fields:" << error;
+                        }
+
+                        const auto refresh_token = data.find(QStringLiteral("refresh_token"));
+                        if (refresh_token != data.constEnd()) {
+                            newRefreshToken = refresh_token.value().toString();
+                        }
+                    }
+                }
+                Q_EMIT refreshFinished(accessToken, newRefreshToken);
+            });
+        };
+        connect(this, &OAuth::fetchWellKnownFinished, this, [this, refresh] {
+            if (_registrationEndpoint.isValid()) {
+                auto registerJob = new RegisterClientJob(_networkAccessManager, _dynamicRegistrationData, _registrationEndpoint, this);
+                connect(registerJob, &RegisterClientJob::finished, this, [this, refresh](const QString &clientId, const QString &clientSecret) {
+                    _clientId = clientId;
+                    _clientSecret = clientSecret;
+                    refresh();
+                });
+                connect(registerJob, &RegisterClientJob::errorOccured, this, [this](const QString &error) {
+                    qCWarning(lcOauth) << "Failed to dynamically register the client, try the default client id" << error;
+                    Q_EMIT refreshFinished(QString(), QString());
+                });
+                registerJob->start();
+            } else {
+                refresh();
+            }
+        });
+        fetchWellKnown();
+    };
+
+    auto credentialsJob = _account->credentialManager()->get(dynamicRegistrationDataC());
+
+    connect(credentialsJob, &CredentialJob::finished, this, [=] {
+        credentialsJob->deleteLater();
+
+        const auto data = [credentialsJob]() -> QVariantMap {
+            if (credentialsJob->data().isValid()) {
+                return credentialsJob->data().value<QVariantMap>();
+            } else {
+                qCCritical(lcOauth) << "Failed to read client id" << credentialsJob->errorString();
+                return {};
+            }
+        }();
+
+        _dynamicRegistrationData = data;
+
+        doRefresh();
```

Comment:
Does using the lambda doRefresh actually solve a problem?

I'm kind of getting the impression that one linear spaghetti style function here would be easier to parse than what we  got here with  doRefresh and the parse lambda.

Reply:
I tried to keep the original function encapsulated. It's already way too complex. But I also don't mind refactoring this, if you wish.